server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java

```diff
@@ -0,0 +1,703 @@
+     * so that we can clean up the batch if it is empty and release the fetching resources
+     *
+     * @param shardRouting shard to be removed
+     */
+    protected void safelyRemoveShardFromBatch(ShardRouting shardRouting) {
+        String batchId = shardRouting.primary() ? getBatchId(shardRouting, true) : getBatchId(shardRouting, false);
+        if (batchId == null) {
+            logger.debug("Shard[{}] is not batched", shardRouting);
+            return;
+        }
+        ConcurrentMap<String, ShardsBatch> batches = shardRouting.primary() ? batchIdToStartedShardBatch : batchIdToStoreShardBatch;
+        ShardsBatch batch = batches.get(batchId);
+        batch.removeFromBatch(shardRouting);
+        deleteBatchIfEmpty(batches, batchId);
+    }
+
+    /**
+     * Safely remove shard from both the batches irrespective of its primary or replica,
+     * For the corresponding shardId. The method intends to clean up the batch if it is empty
+     * after removing the shard
+     * @param shardRouting shard to remove
+     */
+    protected void safelyRemoveShardFromBothBatch(ShardRouting shardRouting) {
+        String primaryBatchId = getBatchId(shardRouting, true);
+        String replicaBatchId = getBatchId(shardRouting, false);
+        if (primaryBatchId == null && replicaBatchId == null) {
+            return;
+        }
+        if (primaryBatchId != null) {
+            ShardsBatch batch = batchIdToStartedShardBatch.get(primaryBatchId);
+            batch.removeFromBatch(shardRouting);
+            deleteBatchIfEmpty(batchIdToStartedShardBatch, primaryBatchId);
+        }
+        if (replicaBatchId != null) {
+            ShardsBatch batch = batchIdToStoreShardBatch.get(replicaBatchId);
+            batch.removeFromBatch(shardRouting);
+            deleteBatchIfEmpty(batchIdToStoreShardBatch, replicaBatchId);
+        }
+    }
+
+    private void deleteBatchIfEmpty(ConcurrentMap<String, ShardsBatch> batches, String batchId) {
+        if (batches.containsKey(batchId)) {
+            ShardsBatch batch = batches.get(batchId);
+            if (batch.getBatchedShards().isEmpty()) {
+                Releasables.close(batch.getAsyncFetcher());
+                batches.remove(batchId);
+            }
+        }
+    }
+
+    protected String getBatchId(ShardRouting shardRouting, boolean primary) {
+        ConcurrentMap<String, ShardsBatch> batches = primary ? batchIdToStartedShardBatch : batchIdToStoreShardBatch;
+
+        return batches.entrySet()
+            .stream()
+            .filter(entry -> entry.getValue().getBatchedShards().contains(shardRouting.shardId()))
+            .findFirst()
+            .map(Map.Entry::getKey)
+            .orElse(null);
+    }
+
+    @Override
+    public AllocateUnassignedDecision explainUnassignedShardAllocation(ShardRouting unassignedShard, RoutingAllocation routingAllocation) {
```

Comment:
If we are adding the fix for allocation explain in this new class only, we can close https://github.com/opensearch-project/OpenSearch/pull/10348. Doing it here only makes sense, rather than raising a new PR for only this change. I will add the test for this change in the same PR as well.

Reply:
The changes in this PR are mainly around Allocation/batching. You can make your PR dependent on my PR and can merge separately